crates/fuel-core/src/service/genesis/task_manager.rs

```diff
@@ -1,74 +1,126 @@
-use std::future::Future;
+use std::{
+    future::Future,
+    sync::Arc,
+};
 
 use fuel_core_services::StateWatcher;
 use futures::{
+    stream::FuturesUnordered,
     StreamExt,
     TryStreamExt,
 };
 use itertools::Itertools;
 use tokio::task::JoinSet;
+use tokio_util::sync::CancellationToken;
 
 pub struct TaskManager<T> {
     set: JoinSet<anyhow::Result<T>>,
-    cancel: CancellationToken,
+    outside_cancel_listener: MultiCancellationToken,
+    cancel_tasks: CancellationToken,
 }
 
-#[cfg_attr(feature = "test-helpers", derive(Default))]
-#[derive(Clone)]
-pub struct CancellationToken {
-    task_cancellator: tokio_util::sync::CancellationToken,
-    state_watcher: StateWatcher,
+#[async_trait::async_trait]
+pub trait NotifyCancel {
+    async fn wait_until_cancelled(&self) -> anyhow::Result<()>;
+    fn is_cancelled(&self) -> bool;
 }
 
-#[cfg(feature = "test-helpers")]
-impl From<tokio_util::sync::CancellationToken> for CancellationToken {
-    fn from(token: tokio_util::sync::CancellationToken) -> Self {
-        Self {
-            task_cancellator: token,
-            ..Default::default()
+#[async_trait::async_trait]
+impl NotifyCancel for CancellationToken {
+    async fn wait_until_cancelled(&self) -> anyhow::Result<()> {
+        self.cancelled().await;
+        Ok(())
+    }
+    fn is_cancelled(&self) -> bool {
+        self.is_cancelled()
+    }
+}
+
+#[async_trait::async_trait]
+impl NotifyCancel for StateWatcher {
+    async fn wait_until_cancelled(&self) -> anyhow::Result<()> {
+        while !self.is_cancelled() {
+            self.clone().changed().await?;
         }
+
+        Ok(())
     }
+
+    fn is_cancelled(&self) -> bool {
+        let state = self.borrow();
+        state.stopping() || state.stopped()
+    }
+}
+
+/// A token that can be used to monitor multiple cancellation sources.
+#[derive(Default, Clone)]
+pub struct MultiCancellationToken {
+    sources: Vec<Arc<dyn NotifyCancel + Send + Sync>>,
 }
 
-impl CancellationToken {
-    #[cfg(test)]
-    pub async fn cancelled(mut self) -> anyhow::Result<()> {
-        tokio::select! {
-            _ = self.task_cancellator.cancelled() => Ok(()),
-            result = self.state_watcher.wait_stopping_or_stopped() => result
+impl MultiCancellationToken {
+    pub fn from_single(source: impl NotifyCancel + Send + Sync + 'static) -> Self {
+        let mut token = Self::default();
+        token.insert(source);
+        token
+    }
+
+    /// Note: Adding a new source to the token will not affect already running futures.
+    pub fn insert(&mut self, source: impl NotifyCancel + Send + Sync + 'static) {
+        self.sources.push(Arc::new(source));
+    }
+}
+
+#[async_trait::async_trait]
+impl NotifyCancel for MultiCancellationToken {
+    async fn wait_until_cancelled(&self) -> anyhow::Result<()> {
```

Comment:
It is not clear what is the purpose of implementing `NotifyChannel` for the `MultiCancellationToken`.

It seems you don't use `MultiCancellationToken::wait_until_cancelled` anywhere except tests=) You only use `is_cancelled` where you select between `StateWatcher::is_cancelled || CancellationToken:: is_cancelled`. Maybe it makes more sense to just create a simple type that has two fields=)

<img width="407" alt="image" src="https://github.com/FuelLabs/fuel-core/assets/18346821/c468c558-6dc5-4536-a853-e507fd9aa6b9">

Reply:
The thought process:
* `TaskManager` needs to be given something to cancel it. In the case of import, cancellation is signaled via the `StateWatcher` being in a `Stopping` or `Stopped` state. For exports you don't have the `StateWatcher`, it felt unnatural to create one force it to be in `Started` and later on have it be in `Stopping` upon receiving a signal. Since we don't have a service for exporting a cancellation token fit best.
* the task manager also has an internal reason to stop -- upon encountering any error. So it needs to register its own way of signaling to stop. Currently that is a cancellation token.
* The tasks need to be given something that signals when either when something on the outside said stop (a `StateWatcher` or a outside `CancellationToken`) or whether the `TaskManager` itself ordered the stop (due to an error). So that means either combine those two (the outside signal and the internal one) into one struct or pass two arguments to each task, which is unsightly and would beg the question "why two sources of cancellation?"

So whatever I replace the `MultiCancellationToken` with needs to have two items:
a) an outside signal (a `StateWatcher` or a `CancellationToken`)
b) an internal signal (a `CancellationToken`)

I couldn't find a name for such a structure and opted for generalizing the problem into a listen to n sources and report if any of them signal cancelled.

Reply:
`CancelationToken` already is good enough=) It is just an internal type for the `TaskManager`

Reply:
I've made it more specific. Lemme know your thoughts.

Reply:
You don't need to implement `NotifyCancel` for `CancellationToken` and you don't need to test how it works=)

Reply:
I need a unified interface because `CancellationToken` is being used by the tasks spawned on `TaskManager`. Either I implement the trait or I just copy the trait implementation into an `impl CancellationToken` block. 

If `NotifyCancel` is a thing, I'd expect it to be implemented on `task_manager::CancellationToken`.

If it has a behavior such that it will signal cancel if either the internal signal or the internal one is cancelled then it needs tests to validate this behavior.

Reply:
https://github.com/FuelLabs/fuel-core/pull/1832/commits/2f023820df86ef7ee207ed950622675425aca78e